### PR TITLE
Revise API of gibbs functions

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR revises the API of discrete_gibbs_fn and subsample_gibbs_fn for both:
+ improve flexibility: currently, a single `gibbs_fn` together with its restricted arguments is not flexible enough to initialize necessary information for the metropolis update other than the HMC (e.g. the behavior `use_unconstrained_gibbs_fn`). In addition, this enables further enhancements such as: passing `potential_energy`, `potential_energy_gen`,... between gibbs_fn and the main kernel `sample` method, and getting other diagnostics rather than the current one in HMCGibbs.
+ clean up the user API: avoid unnecessary users' specifications such as `model`, `model_args`, `model_kwargs`, `gibbs_fn`, `gibbs_sites`.

To do that, I inherited the class `HMCGibbs` (NB: MixedHMC and MixedHMCGibbs in #826 also inherited from this class).

Things are changed in this PR:
+ the api of the above two functions (I probably move things around a bit but the implementation should be the same)
+ add `self._use_constrained_gibbs_fn` to tell `gibbs_fn` if `z_hmc` is constrained or unconstrained
+ add `self._prototype_trace` to avoid recomputing the trace
+ marginalization in `DiscreteHMCGibbs` requires an explicit `infer={"enumerate": "parallel"}` specification in `sample` statement, rather than automatically as before.

cc @OlaRonning I think this change is also more appropriate for HMCECS, where it might need a `rng_key` to initialize some thing (I seem to me that proxy stuff there is sensitive to the seed of the prototype trace)

### CODE REVIEW

**Review by OlaRonning — COMMENTED** [review]
Hi @fehiepsi, this looks good! 

The proxies should not have any random behavior when initialized, however, the model wrapper will depend on the estimator and this will be simpler if it's just set at `init`.

Have a couple of comments, see above.

File: numpyro/infer/hmc_gibbs.py [hunk 1]
```diff
@@ -315,12 +264,9 @@ def discrete_gibbs_fn(model, model_args=(), model_kwargs={}, *, random_walk=Fals
...
 
     def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
-        # convert to unconstrained values
-        z_hmc = {k: biject_to(prototype_trace[k]["fn"].support).inv(v)
-                 for k, v in hmc_sites.items()
-                 if k in prototype_trace and prototype_trace[k]["type"] == "sample"}
+        z_hmc = hmc_sites
         use_enum = len(set(support_sizes) - set(gibbs_sites)) > 0
-        wrapped_model = _wrap_model(model)
+        wrapped_model = model
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/hmc_gibbs.py:269]: Good point! I'll make the change.

File: numpyro/infer/hmc_gibbs.py [hunk 2]
```diff
@@ -93,6 +92,8 @@ def __init__(self, inner_kernel, gibbs_fn, gibbs_sites):
         self.inner_kernel._model = _wrap_model(inner_kernel.model)
         self._gibbs_sites = gibbs_sites
         self._gibbs_fn = gibbs_fn
+        self._use_constrained_gibbs_fn = True
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/hmc_gibbs.py:95]: I think we can just use the default behavior, to not populate the api, until users need the `False` pattern.

File: numpyro/infer/hmc_gibbs.py [hunk 3]
```diff
@@ -359,35 +307,33 @@ def body_fn(i, val):
...
-    hence might incur a high variance.
+    A subclass of :class:`HMCGibbs` which performs Metropolis updates for discrete latent sites.
 
-    The function can partition named subsample statements and update only one block in the parition
-    to improve acceptance rate of proposed subsamples as detailed in [3].
-    .. note:: New subsample indices are proposed randomly with replacement at each MCMC step.
+    .. note:: The site update order is randomly permuted at each step.
 
-    **References:**
+    .. note:: This class supports mixing with enumeration. To marginalize out a discrete
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:318]: nit: This class supports enumeration of discrete latent variables.

File: numpyro/infer/hmc_gibbs.py [hunk 4]
```diff
@@ -397,35 +343,42 @@ def subsample_gibbs_fn(model, model_args=(), model_kwargs={}, *, num_blocks=1):
...
-        >>> gibbs_fn = subsample_gibbs_fn(model, (data,), num_blocks={'N': 10})
-        >>> kernel = HMCGibbs(NUTS(model), gibbs_fn, gibbs_sites=["N"])
-        >>> mcmc = MCMC(kernel, 1000, 1000)
-        >>> mcmc.run(random.PRNGKey(0), data)
-        >>> samples = mcmc.get_samples()["x"]
-        >>> assert abs(jnp.mean(samples).copy() - 1.) < 0.1
+        >>> probs = jnp.array([0.15, 0.3, 0.3, 0.25])
+        >>> locs = jnp.array([-2, 0, 2, 4])
+        >>> kernel = DiscreteHMCGibbs(NUTS(model), modified=True)
+        >>> mcmc = MCMC(kernel, 1000, 100000, progress_bar=False)
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:355]: why do this many samples if there's no `assert`?

File: numpyro/infer/hmc_gibbs.py [hunk 5]
```diff
@@ -441,13 +394,90 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
...
     return gibbs_fn
+
+
+class HMCECS(HMCGibbs):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    HMC with Energy Conserving Subsampling.
+
+    A subclass of :class:`HMCGibbs` to perform HMC-within-Gibbs for model with subsampling
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:415]: nit: for performing

Comment (martinjankowiak) [reply numpyro/infer/hmc_gibbs.py:415]: nit: for models with subsample statements using the

File: numpyro/infer/hmc_gibbs.py [hunk 6]
```diff
@@ -441,13 +394,90 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
...
+    statements using :class:`~numpyro.plate` primitive. This implements the Algorithm 1
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:416]: nit: This implements Algorithm 1

File: numpyro/infer/hmc_gibbs.py [hunk 7]
```diff
@@ -441,13 +394,90 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
...
+    of reference [1] but uses a naive estimation (without control variates) of log likelihood,
+    hence might incur a high variance.
+
+    The function can partition named subsample statements and update only one block in the parition
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:420]: typo: partition

File: numpyro/infer/hmc_gibbs.py [hunk 8]
```diff
@@ -441,13 +394,90 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
...
+    to improve acceptance rate of proposed subsamples as detailed in [3].
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:421]: nit: improve the acceptance rate

File: numpyro/infer/hmc_gibbs.py [hunk 9]
```diff
@@ -441,13 +394,90 @@ def gibbs_fn(rng_key, gibbs_sites, hmc_sites):
...
+        ...     with numpyro.plate("N", data.shape[0], subsample_size=100):
+        ...         batch = numpyro.subsample(data, event_dim=0)
+        ...         numpyro.sample("obs", dist.Normal(x, 1), obs=batch)
+        ...
+        >>> data = random.normal(random.PRNGKey(0), (10000,)) + 1
+        >>> kernel = HMCECS(NUTS(model), num_blocks=10)
+        >>> mcmc = MCMC(kernel, 1000, 1000)
+        >>> mcmc.run(random.PRNGKey(0), data)
+        >>> samples = mcmc.get_samples()["x"]
+        >>> assert abs(jnp.mean(samples).copy() - 1.) < 0.1
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_gibbs.py:458]: why copy?

File: numpyro/infer/hmc_gibbs.py (hunk 4 above)
Comment (fehiepsi) [reply numpyro/infer/hmc_gibbs.py:355]: Oops, probably this losts during a commit. I'll reverse the assertations.

File: numpyro/infer/hmc_gibbs.py (hunk 9 above)
Comment (fehiepsi) [reply numpyro/infer/hmc_gibbs.py:458]: There is no reason. I'll remove it.

**Comment by fehiepsi** [comment]
Thanks for reviewing, @OlaRonning and @martinjankowiak!

**Review by martinjankowiak — APPROVED** [review]
